scripts/rewards_simple.py

```diff
@@ -0,0 +1,125 @@
+        self.currency = currency
+
+
+class Rewards:
+    def __init__(self):
+        self.total_stake = 0
+        self.reward_per_token = 0
+        self.stake = {}
+        self.reward_tally = {}
+
+    def deposit_stake(self, address, amount):
+        if address not in self.stake:
+            self.stake[address] = 0
+            self.reward_tally[address] = 0
+
+        self.stake[address] = self.stake[address] + amount
+        self.reward_tally[address] = self.reward_tally[address] + \
+            self.reward_per_token * amount
+        self.total_stake = self.total_stake + amount
+
+    def distribute(self, reward):
+        if reward == 0:
+            return
+        if self.total_stake == 0:
+            raise Exception("Invalid total stake")
+
+        self.reward_per_token = self.reward_per_token + reward / self.total_stake
+
+    def compute_reward(self, address):
+        return self.stake[address] * self.reward_per_token - self.reward_tally[address]
+
+    def withdraw_stake(self, address, amount):
+        if address not in self.stake:
+            raise Exception("Stake not found")
+
+        if amount > self.stake[address]:
+            raise Exception("Amount is too high")
+
+        self.stake[address] = self.stake[address] - amount
+        self.reward_tally[address] = self.reward_tally[address] - \
+            self.reward_per_token * amount
+        self.total_stake = self.total_stake - amount
+        return amount
+
+    def withdraw_reward(self, address):
+        reward = self.compute_reward(address)
+        self.reward_tally[address] = self.stake[address] * \
+            self.reward_per_token
+        return reward
+
+    def update_stake(self, address, amount):
+        if amount > 0:
+            self.deposit_stake(address, amount)
+        else:
+            self.withdraw_stake(address, abs(amount))
+
+
+class VaultRegistry:
+    def __init__(self):
+        self.exchange_rate = {}
+        self.secure_threshold = {}
+
+        self.collateral = {}
+        self.vault_rewards = Rewards()
+
+    # relative to btc
+    def set_exchange_rate(self, currency, value):
+        self.exchange_rate[currency] = value
+
+    def set_secure_threshold(self, currency, value):
+        self.secure_threshold[currency] = value
+
+    def deposit_collateral(self, vault, amount):
+        if vault not in self.collateral:
+            self.collateral[vault] = 0
+
+        self.collateral[vault] = self.collateral[vault] + amount
+
+        capacity = 0
+        if vault in self.vault_rewards.stake:
+            capacity = self.vault_rewards.stake[vault]
+
+        capacity_delta = self.vault_capacity(vault) - capacity
+        self.vault_rewards.update_stake(vault, capacity_delta)
+
+    def vault_capacity(self, vault):
+        return self.collateral[vault] \
+            / self.exchange_rate[vault.currency] \
+            / self.secure_threshold[vault.currency]
+
+    def distribute(self, reward):
```

Comment:
Something is wrong here. Where it gets called second time, it adds the reward on top of the value from the first call.
It also seems to split the rewards based on the previous capacity, not the updated one based on new capacity.

Reply:
> Something is wrong here. Where it gets called second time, it adds the reward on top of the value from the first call.

It should, shouldn't it? If you distribute 5 kint, and then another 10 kint, you want the total distributed rewards to be 15 kint? Can you clarify what you think the problem is?

> It also seems to split the rewards based on the previous capacity, not the updated one based on new capacity.

Distributing rewards does not change capacity, for two reasons. First, the rewards are not actually received until you explicitly withdraw rewards. Second, rewards that you withdraw are not automatically added as collateral. In general the reward currency might not even be the same currency as the collateral currency.

Reply:
Okay makes sense. I was somehow inferring that this would include a 'withdraw' of distributions.

The latter point was directed to a test I made where I changed the price or custom threshold for a vault and distributed rewards based on that new capacity. However, the rewards got distributed without including these changes (same ratio as before).